test/diameter_avp_filter_SUITE.erl

```diff
@@ -0,0 +1,403 @@
+%%% Common Test callbacks
+%%%===================================================================
+
+all() -> [default_filter,
+	  simple_root,
+	  simple_path,
+	  conditional_instance,
+	  ignore_single_condition_path].
+
+init_per_suite(Config0) ->
+    Config = [{handler_under_test, ?HUT} | Config0],
+    application:load(ergw_aaa),
+    [application:set_env(ergw_aaa, Key, Opts)
+     || {Key, Opts} <- ?CONFIG],
+    init_test_info_ets(),
+    meck_init(Config),
+    TestTransports = [[{transport_module, diameter_tcp},
+		       {capabilities, [{'Origin-Host', Host}]},
+		       {transport_config, [{reuseaddr, true}, {ip, IP}]}]
+		      || {IP, Host} <- ?TEST_SERVER_TRANSPORTS],
+    diameter_test_server:start(?TEST_SERVER_CALLBACK_OVERRIDE,
+			       TestTransports),
+    {ok, _} = application:ensure_all_started(ergw_aaa),
+    case wait_for_diameter(?SERVICE, 10) of
+      ok -> Config;
+      Other -> end_per_suite(Config), {skip, Other}
+    end.
+
+end_per_suite(Config) ->
+    meck_unload(Config),
+    application:stop(ergw_aaa),
+    application:unload(ergw_aaa),
+    diameter_test_server:stop(),
+    ok.
+
+init_per_testcase(Config) -> meck_reset(Config), Config.
+
+end_per_testcase(Config) -> 
+    meck_validate(Config),
+    ok.
+
+%%%===================================================================
+%%% Test cases
+%%%
+%%% The simplest way to test this is to fire 2 requests, 1 without
+%%% and another with the filtering we test, and compare the result.
+%%%===================================================================
+
+%%====================================================================
+%% Default filter (for ro it is ['3GPP-IMSI'])
+%% override default filter with empty list and check if before we have
+%% '3GPP-IMSI' filtered out by the default filter, then present after
+%% empty filter is applied
+%%====================================================================
+default_filter(_Config) ->
+    {Before, After} = do_filter_test([]),
+    ?equal(undefined, maps:get('3GPP-IMSI', Before, undefined)),
+    ?match([_], maps:get('3GPP-IMSI', After)),
+    ok.
+
+%%====================================================================
+%% Simple filter of single AVPs 'User-Name' from the root of
+%% the AVP tree as well as the '3GPP-IMSI' as the filter overrides the
+%% handler default
+%%====================================================================
+simple_root(_Config) ->
+    {Before, After} = do_filter_test([['3GPP-IMSI'], ['User-Name']]),
+
+    ?equal(undefined, maps:get('3GPP-IMSI', Before, undefined)),
+    ?match([_], maps:get('User-Name', Before, undefined)),
+    
+    ?equal(undefined, maps:get('3GPP-IMSI', After, undefined)),
+    ?equal(undefined, maps:get('User-Name', After, undefined)),
+    ok.
+
+%%====================================================================
+%% Simple path filter to remove 
+%% 'Service-Information' -> 'PS-Information' -> '3GPP-MS-TimeZone'
+%% and '3GPP-IMSI'
```

Comment:
Why filter 3GPP-IMSI repeatedly? There is probably a good reason, but I do not understand.

Reply:
The reason for this is that 3GPP-IMSI (and the other 'extra AVPS' in the table in the description) is part of the default filters that are intended to make the deployments without this new filter configuration backward compatible. Filter configuration overrides the default, therefore those AVPs need to be included there, unless the intention is to send them. I'll see it I can improve the description.